✨ feat:  added form I

acapolungo opened the pull request:
Ajout de la forme I

acapolungo committed:
✨ feat:  added form I

acapolungo committed:
✨ feat: added rotation Action in reducer

addition of the rotation function with in the coordinates of the shapes a switch for the position of each shape. Our check functions now include this position and we can rotate the shape with the spacebar

Issue #21

dbesserman commented on src/types.tsx:
```diff
@@ -2,18 +2,19 @@
 export type Grid = CellType[][];
 export type Color = "yellow" | "orange" | "purple" | "blue" | "red" | "grey";
+export type Shape = "i" | "o";
+// export type Shape = "i" | "j" | "l" | "o" | "s" | "t" | "z";
```
relis toi

dbesserman commented on src/types.tsx:
```diff
@@ -2,18 +2,19 @@
 export type CellType = { color?: Color; isActive?: boolean; isEmpty: boolean; status?: Status };
 export type Grid = CellType[][];
 export type Color = "yellow" | "orange" | "purple" | "blue" | "red" | "grey";
+export type Shape = "i" | "o";
```
Il faudrait faire des types un peu plus complexes. Une forme devrait être représenté par un objet avec comme propriété
shapeType qui peut pour le moment être "i" ou "o"
vectors qui est un array d'arrays de vecteurs
currentPositionIndex qui est un nombre entier (index qui permet de retrouver l'array de vecteurs actuel ou suivant selon le besoin)

acapolungo replied:
Je n'ai pas trop compris la demande, tu veux que dans le type shapeType j'introduise le vecteur de la forme également ainsi qu'une currentPositionIndex ?

du genre :

```typescript
shapeType = 
letter: "i" | "o";
currentPositionIndex : Position;
currentShapeVectors: Vector[];
```

dbesserman replied:
oui sauf que je mettrais number comme type pour `currentPositionIndex` parce que ce nombre est variable en fonction de la forme. avoir une position de 4 pour "o" n'a pas de sens.

dbesserman commented on src/App.tsx:
```diff
@@ -3,14 +3,21 @@ import './App.css';
     return color
 };
 
+// const arrayOfShapes: Color[] = ["i", "j", "l", "o", "s", "t", "z"];
```
relis toi. Çà devrait être acquis maintenant qu'avant de pousser du code tu te relis..

acapolungo replied:
oui.

dbesserman commented on src/App.tsx:
```diff
@@ -3,14 +3,21 @@ import './App.css';
 };
 
+// const arrayOfShapes: Color[] = ["i", "j", "l", "o", "s", "t", "z"];
+const randomShapes = (): Shape => {
```
pourquoi randomShapes est au pluriel?

dbesserman commented on src/App.tsx:
```diff
@@ -3,14 +3,21 @@ import './App.css';
+// const arrayOfShapes: Color[] = ["i", "j", "l", "o", "s", "t", "z"];
+const randomShapes = (): Shape => {
+    const arrayOfShapes: Shape[] = ["o", "i"];
+    const shape = arrayOfShapes[Math.floor(Math.random() * arrayOfShapes.length)]
```
pas besoin de créer de variable shape si tu le retourne juste après.
Tu peux directement faire
```js
return arrayOfShapes[Math.floor(Math.random() * arrayOfShapes.length)]
```

dbesserman commented on src/App.tsx:
```diff
@@ -69,33 +80,36 @@ const tetrisGridCopy = (tetrisGrid: Grid): Grid => tetrisGrid.map(row => [...row
         case 'TRY_INTRODUCE_SHAPE':
 
             const activeCurrentCell: CellType = { color: nextShapeColor, isActive: true, isEmpty: false }
+            const introduceShapeCellsCoordinates = cellCoordinates(initialState.referenceCellCoordinate, shapesVectors(nextShape, 1))
```
la variable s'appelle "introduire les coordonnées de la cellule de la forme". C'est vraiment comme çà que tu voulais l'appeler? Il faut que tu prennes l'habitude à chaque fois que tu nommes une variable de passer par google traduction

dbesserman commented on src/App.tsx:
```diff
@@ -164,14 +176,42 @@ const reducer = (state: TetrisState, action: Action): TetrisState => {
+        case 'TRY_ROTATE_SHAPE':
+            let newPosition;
+            newPosition = position
+            newPosition++
```
`++` crée une mutation donc il vaut mieux l'éviter en React.
Privilégie
```js
newPosition = position + 1
```

dbesserman commented on src/App.tsx:
```diff
@@ -164,14 +176,42 @@ const reducer = (state: TetrisState, action: Action): TetrisState => {
+                    position: newPosition as Position,
+                }
+            } else {
+                // fallbackCallback()
```
?

acapolungo replied:
Le même souci avec le fallbackcallback, le type est demandé dans les fonction qui ne l'ont pas, et est block scopé si je le lance dans deux action.
Je n'arrive pas à le remonter comme tu l'as fais avec  clearCurrentTimeout, il me créer des erreurs.

dbesserman replied:
Donc tu veux avoir le fallback callback mais tu as des soucis avec typescript? C'est çà?

dbesserman commented on src/App.tsx:
```diff
@@ -29,7 +36,11 @@ const initialState: TetrisState = {
+    firstRenderHappened: false,
+    currentShape: randomShapes(),
+    nextShape: randomShapes(),
+    shapeVector: [],
```
1. une forme a un seul vecteur?
2. comment se réfère t'on à la forme en question?

dbesserman commented on src/App.tsx:
```diff
@@ -183,35 +223,69 @@ const tetrisGridWithoutShapeApplied = (tetrisGrid: Grid, referenceCellCoordinate
 }
 
-const tetrisGridWithShapeMoved = (tetrisGrid: Grid, referenceCellCoordinate: Coordinate, activeCell: CellType, direction: Direction): Grid => {
+const tetrisGridWithShapeMoved = (tetrisGrid: Grid, referenceCellCoordinate: Coordinate, activeCell: CellType, direction: Direction, shape: Vector[]): Grid => {
```
l'attribut s'appelle "forme" mais c'est un array de vectors
préfère shapeVectors ou currentShapeVectors

acapolungo replied:
Oui parce que elle renvoit la grille totale avec la forme après rotation.

acapolungo replied:
remplacé par tetrisGridAfterShapeRotation ! / tetris grid après rotation de la forme

dbesserman replied:
top

dbesserman commented on src/App.tsx:
```diff
@@ -183,35 +223,69 @@ const tetrisGridWithoutShapeApplied = (tetrisGrid: Grid, referenceCellCoordinate
+    return tetrisGridWithShapeApplied(
+        tetrisGridWithoutShapeApplied(gridCopy, referenceCellCoordinate, shape),
+        referenceCellCoordinate,
+        newShape,
```
new ne donne pas beaucoup d'info sur ce qui s'est passé.
rotatedShapeVectors est plus parlant par exemple

dbesserman commented on src/App.tsx:
```diff
@@ -271,14 +345,22 @@ const moveDownIsPossible = (grid: Grid, activeCellsCoordinates: Coordinate[]): b
+    referenceCellsCoordinates.shift()
+    return referenceCellsCoordinates.every((referenceCellCoordinates) => {
+        const [rowIndex, columnIndex] = referenceCellCoordinates
+        return cellIsEmpty(grid, rowIndex, columnIndex);
```
que se passe t'il si les coordonnées de la forme après rotation sont en dehors du tetrisGrid

acapolungo replied:
ça plante effectivement, merci.

dbesserman commented on src/App.tsx:
```diff
@@ -271,14 +345,22 @@ const moveDownIsPossible = (grid: Grid, activeCellsCoordinates: Coordinate[]): b
     })
 }
 
+const shapeCanBeRotated = (grid: Grid, referenceCellsCoordinates: Coordinate[]): boolean => {
```
Il n'y a qu'une cellule de reférence

dbesserman commented on src/App.tsx:
```diff
@@ -271,14 +345,22 @@ const moveDownIsPossible = (grid: Grid, activeCellsCoordinates: Coordinate[]): b
 }
 
+const shapeCanBeRotated = (grid: Grid, referenceCellsCoordinates: Coordinate[]): boolean => {
+    referenceCellsCoordinates.shift()
```
shift est une action destructrice. Il vaut mieux éviter en React.

acapolungo replied:
Oui justement j'ai besoin de couper la première cellule pour qu'elle ne soit pas testée, sinon comme elle est active et donc plaine le test ne passera jamais à cause de cette case.

Donc j'avais eu comme idée de shift la coordonnée de la case.

dbesserman replied:
regarde ce que tu as fais ligne 333 pour la fonction  `cellBelowIsFree`

dbesserman commented on src/components/NextShapePreview.tsx:
```diff
@@ -1,37 +1,64 @@
+  return new Array(6).fill(rowsOfCells);
+}
+
+const landmark = (shape: Shape): Coordinate => {
```
Je suis pas sur que çà soit nécessaire. Je pense que çà complique un peu les chose.
Prenons un i horizontal. si la cellule de référence est la cellule la plus a gauche, les vecteurs seront:
```js
[[0,0], [0,1], [0,2], [0, 3]]
```
On peut aussi se dire que la cellule de référence est la deuxième case. Dans ce cas, il y'a juste à changer les vecteurs.
```js
[[0, -1], [0, 0], [0, 1], [0, 2]]
```
Je pense que tu peux te passer de la notion de `landmark`. À voir

acapolungo replied:
C'est un point de repère pour placer la case dans NewShapePreview, à savoir une cordonnée spécifique selon la forme pour placer manuellement la présentation de la forme à venir.

dbesserman replied:
On en a discuté samedi. Il vaut mieux avoir une propriété `previewReferenceCellCoordinate` pour chaque forme

dbesserman reviewed (commented):
Il y'a du mieux et des regressions.
Pour ce qui de la manière de la manière de modéliser les formes, je pense qu'on peut améliorer, mais tu as trouvé une solution correcte par toi même. C'est bien

acapolungo committed:
🐞 fix: fix app crash when shape rotates

crash resolution with a test that checks if the cells are not undefined. Changed the name of the shape types. Updated the name of some functions for the overall understanding of the code

#21

acapolungo committed:
🦄 refactor: Slightly more complex types for shapes

Added letter postion and vector in for each shape. Changing types for each shape element.

acapolungo committed:
🦄 refactor: adding a preview coordinate in our shape

We have coordinates in our shapes which we pass to the nextShape component in order to display the shape in the right place of the preview.

Issue #21

dbesserman commented on src/App.tsx:
```diff
@@ -12,7 +12,7 @@ const randomColor = (): Color => {
 };
 
 const randomShape = (): shapeType => {
```
un type commence par une majuscule (PascalCase)

dbesserman commented on src/App.tsx:
```diff
@@ -222,9 +222,9 @@ const tetrisGridWithoutShapeApplied = (tetrisGrid: Grid, referenceCellCoordinate
 }
 
-const shapesVectors = (shape: shapeType, position: Position): Vector[] => {
+const shapesVectors = (shape: shapeName, currentPositionIndex: number): Vector[] => {
```
tu utilises `shape`, `shapeName`, et `currentLetter` pour désigner la même chose.
`shape` est confusant parce que tu t'attends à que ce soit un objet de type `ShapeType`
`currentLetter` n'est pas clair. Pourquoi current? on utilise current pour distinguer la forme actuelle (currentShape) de la forme suivante (nextShape). Pourtant la forme suivante (`nextShape`) a une lettre actuelle (`currentLetter`).
Une solution est à l'intérieur de l'objet d'avoir une propriété `name` afin de pouvoir appeller
```js
currentShape.name
```
et à l'extérieur de la forme avoir `shapeName`. Par exemple ici tu aurais
```js
const shapesVectors = (shapeName: shapeName, currentPositionIndex: number): Vector[] => {
```

acapolungo replied:
Oui je vais renommer.

dbesserman commented on src/types.tsx:
```diff
@@ -2,11 +2,16 @@
 export type Grid = CellType[][];
 export type Color = "yellow" | "orange" | "purple" | "blue" | "red" | "grey";
-export type shapeType = "i" | "o";
+// export type shapeType = "i" | "o";
```
relis toi

dbesserman commented on src/types.tsx:
```diff
@@ -22,8 +27,8 @@ export type TetrisState = {
     firstRenderHappened: Boolean;
-    currentShape: shapeType;
-    nextShape: shapeType;
+    currentShape: shapeName;
```
currentShape et nextShape devraient être de type `ShapeType`

acapolungo replied:
oui en PascalCase

dbesserman commented on src/App.tsx:
```diff
@@ -222,9 +222,9 @@ const tetrisGridWithoutShapeApplied = (tetrisGrid: Grid, referenceCellCoordinate
 
-const shapesVectors = (shape: shapeType, position: Position): Vector[] => {
+const shapesVectors = (shape: shapeName, currentPositionIndex: number): Vector[] => {
     if (shape === 'o') {
```
tu ne devrais pas avoir de `if` ici les informations contenues dans les objets de type `ShapeType` devraient suffire.

dbesserman commented on src/App.tsx:
```diff
@@ -347,18 +347,20 @@ const cellsAreEmpty = (grid: Grid, rowIndex: number, columnIndex: number): boole
+    const [, ...restCoordinatesWithoutFirstCell] = cellsCoordinates;
+
+    return restCoordinatesWithoutFirstCell.every((cellCoordinates) => {
+        const [rowIndex, columnIndex] = cellCoordinates
```
Pourquoi utiliser le pluriel  (`cellsExist` , `cellsAreEmpty`) ? À chaque fois tu fais le test sur plusieurs cellules?
Il fallait just que faire comme tu as déjà fait un peu plus bas
```js
const leftCellExist = (grid: Grid, rowIndex: number, columnIndex: number): boolean => {
    return Boolean(grid[rowIndex][columnIndex - 1] !== undefined);
}
const leftCellIsFree = (grid: Grid, rowIndex: number, columnIndex: number): boolean | undefined => {
    return grid[rowIndex][columnIndex - 1].isEmpty || grid[rowIndex][columnIndex - 1].isActive;
};
```
Et comme çà t'as pas besoin d'enlever le premier élément de l'array

acapolungo replied:
je ne comprends pas vraiment la logique,

Nous passons les coordonnée active d'une forme, et nous tournons la forme à partir d'une case de référence.
Je simule une case active selon son vecteur au moment ou elle tourne
ensuite je passe nos coordonnées dans une fonction qui va vérifier "à partir de la case de référence" si dans notre grille nous avons bien toutes des cases vide de manière à pouvoir tourner la forme.

Si on ne supprime pas la première case (case de référence), le test dans la fonction cellIsEmpty() ne passera jamais

La première case sera toujours active peu importe la forme qu'on passe, la case de référence est la première dans notre cas.

acapolungo replied:
Ok je viens de comprendre, la case est libre ou active !

dbesserman commented on src/components/NextShapePreview.tsx:
```diff
@@ -1,13 +1,13 @@
 }
 
-const landmark = (shape: shapeType): Coordinate => {
+const landmark = (shape: shapeName): Coordinate => {
```
On a dit quand on s'était vu qu'on enlevait les landmarks. Il faut à la place que l'objet de type ShapeType ait une propriété qui s'appelle `previewReferenceCellCoordinate`

acapolungo replied:
je comprend pas mon code est à jour pourtant
tu as bien la dernière version ? je n'ai plus de landmark et previewReferenceCellCoordinate est bien là

acapolungo replied:
```javascript

type PropType = {
  nextShapeColor: Color;
  nextShape: shapeType;
  nextShapeVectors: Vector[]
}

export default function NextShape({ nextShapeColor, nextShape, nextShapeVectors }: PropType) {

  const [nextShapeGrid, updateNextShapeGrid] = useState(emptyNextShapeGrid());
  const { previewReferenceCellCoordinate } = nextShape

  useEffect(() => {
    updateNextShapeGrid(arr => {
      const activeCell = { color: nextShapeColor, isActive: true, isEmpty: false }
      const nextShapeCellsCoordinates = cellCoordinates(previewReferenceCellCoordinate, nextShapeVectors)
      return arr = nextShapeGridWithIntroducedShape(emptyNextShapeGrid(), nextShapeCellsCoordinates, activeCell);
    }
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [nextShape])
```

acapolungo committed:
🦄 refactor: change switch for the shape and tests to rotate

changed the way the witch works without ifs. Change whose box we test to make it rotate. Shape naming change.

#21

acapolungo committed:
🦄 refactor:   complexification of the randomShape shape object

position Vectors and position index for the rotation is in the randomShape object which avoids ifs in ifs in the reducer

#21

dbesserman commented on src/App.tsx:
```diff
@@ -3,14 +3,35 @@ import './App.css';
 import Cell from './components/Cell';
 
 import NextShapePreview from './components/NextShapePreview';
-import { CellType, Color, TetrisState, Action, Source, Coordinate, Vector, Grid, Direction } from './types';
```
coucou
